Source/WebCore/css/StyleProperties.cpp

```diff
@@ -997,9 +997,58 @@ String StyleProperties::getGridValue() const
     return getGridShorthandValue(gridShorthand());
 }
 
+static bool isCustomIdentValue(const CSSValue& value)
+{
+    return is<CSSPrimitiveValue>(value) && downcast<CSSPrimitiveValue>(value).isCustomIdent();
+}
+
 String StyleProperties::getGridShorthandValue(const StylePropertyShorthand& shorthand) const
 {
-    return getShorthandValue(shorthand, " / ");
+    // Look backwards through the longhands to see if there are any trailing
+    // values we can omit.
+    unsigned last = shorthand.length() - 1;
+    for (; last > 0; --last) {
+        auto property = shorthand.properties()[last];
+        auto value = getPropertyCSSValue(property);
```

Comment:
I worry this is using `getPropertyCSSValue` too many times, which can be slow if the vector of properties is big.
I would prefer to call it only once per longhand at the beginning, and then use some logic like Firefox:

```rust
    pub fn can_omit(&self, other: &Self) -> bool {
        if self.is_ident_only() {
            self == other
        } else {
            other.is_auto()
        }
    }
```

```rust
        fn to_css<W>(&self, dest: &mut CssWriter<W>) -> fmt::Result where W: fmt::Write {
            self.grid_row_start.to_css(dest)?;
            let mut trailing_values = 3;
            if self.grid_column_start.can_omit(self.grid_column_end) {
                trailing_values -= 1;
                if self.grid_row_start.can_omit(self.grid_row_end) {
                    trailing_values -= 1;
                    if self.grid_row_start.can_omit(self.grid_column_start) {
                        trailing_values -= 1;
                    }
                }
            }
            let values = [&self.grid_column_start, &self.grid_row_end, &self.grid_column_end];
            for value in values.iter().take(trailing_values) {
                dest.write_str(" / ")?;
                value.to_css(dest)?;
            }
            Ok(())
        }
```

Reply:
Alright, I can do that! I can't think of a nice way to handle all 3 properties with a single function with that approach, but probably not a big deal to duplicate.